### PR TITLE
ESE: Introduce the Enforce Single Exit pass

### PR DESCRIPTION
**Pull request by andrealinux1**

### CODE REVIEW

File: lib/RestructureCFG/EnforceSingleExitPass.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    // We should have collected at least one root
+    revng_assert(Roots.size() != 0);
+
+    // 3: We add an additional `sink_block` where to redirect all the roots
+    //    found only if we have more than one root
+    if (Roots.size() <= 1) {
+      return false;
+    }
```
Comment (pfez) [inline lib/RestructureCFG/EnforceSingleExitPass.cpp:109]: The new block should be added in all the cases where all detected exits are non-trivial.

In all the other cases, there's at least one eligible trivial exits that gets marked as "one true exit".

File: lib/RestructureCFG/EnforceSingleExitPass.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    // 6: Connect all the exit blocks to the newly created `sink_block` with a
+    //    `scope_closer` edge
+    for (BasicBlock *Root : Roots) {
+
+      // Build a `scope_closer` edge from each identified exit block to the
+      // `sink_block`
+      ScopeGraphBuilder SGBuilder(Root->getParent());
+      SGBuilder.addScopeCloser(Root, SinkBlock);
+    }
```
Comment (pfez) [inline lib/RestructureCFG/EnforceSingleExitPass.cpp:143]: With the new logic I suggested at point 3, this should add scope closer edges only from exits that are not the elected one-true-exit.

**Review by pfez — CHANGES_REQUESTED** [review]

File: lib/RestructureCFG/EnforceSingleExitPass.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+/// blocks, and infinite loops too.
+class EnforceSingleExitPassImpl {
+  Function &F;
+
+public:
+  EnforceSingleExitPassImpl(Function &F) : F(F) {}
+
+public:
+  bool run() {
+    llvm::SmallVector<BasicBlock *> Roots;
```
[the remark this answers is not in the document]
Comment (andrealinux1) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: yeah, I used `Root` in order to mimic the terminology used by the LLVM PDT, but I changed this all over the pass to `Exit`, in order to reflect the name itself of the pass `Enforce Single Exit`

File: lib/RestructureCFG/EnforceSingleExitPass.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            HasExit = true;
+            break;
+          }
+        }
+
+        // We are not interested in SCCs that already have an exit, also on the
+        // `ScopeGraph`
+        if (HasExit)
+          break;
+      }
```
[the remark this answers is not in the document]
Comment (andrealinux1) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: I outlined this out here https://github.com/revng/revng/blob/98f7b2d8804ed56f760546885bbbff6c72933843/lib/RestructureCFG/EnforceSingleExitPass.cpp#L23

File: lib/RestructureCFG/EnforceSingleExitPass.cpp (hunk 1 above)
Comment (andrealinux1) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: Ok, I adopted the new strategy that you described, there are a couple of changes around due to this, like keeping two `vector`s for the _trivial_ and _non trivial_ exits https://github.com/revng/revng/blob/98f7b2d8804ed56f760546885bbbff6c72933843/lib/RestructureCFG/EnforceSingleExitPass.cpp#L69, let me know if you approve them

File: lib/RestructureCFG/EnforceSingleExitPass.cpp (hunk 2 above)
Comment (andrealinux1) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: Yeah, I changed a bit the code here, have a look at point 7 https://github.com/revng/revng/blob/98f7b2d8804ed56f760546885bbbff6c72933843/lib/RestructureCFG/EnforceSingleExitPass.cpp#L160 and 8 https://github.com/revng/revng/blob/98f7b2d8804ed56f760546885bbbff6c72933843/lib/RestructureCFG/EnforceSingleExitPass.cpp#L172. I preferred not to mix up the logic of adding the _scope_closer_ for the _trivial_ and _non trivial_ exits, there are a couple of factorization opportunity, but I preferred super clarity here

File: lib/RestructureCFG/EnforceSingleExitPass.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (pfez) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: And just to be super pedantic: `isInfiniteLoop` should take the scc iterator as argument.

File: lib/RestructureCFG/EnforceSingleExitPass.cpp (hunk 1 above)
Comment (pfez) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: Yeah it makes sense.

File: lib/RestructureCFG/EnforceSingleExitPass.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrealinux1) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: We now take the iterator here https://github.com/revng/revng/blob/1a35abb5e36870643588163d5b6f0c0c72feb701/lib/RestructureCFG/EnforceSingleExitPass.cpp#L24
 Regarding the early returns, added them here https://github.com/revng/revng/blob/1a35abb5e36870643588163d5b6f0c0c72feb701/lib/RestructureCFG/EnforceSingleExitPass.cpp#L39 and here https://github.com/revng/revng/blob/1a35abb5e36870643588163d5b6f0c0c72feb701/lib/RestructureCFG/EnforceSingleExitPass.cpp#L44 (I guess you meant to return `true` if you reach the end of the function

File: lib/RestructureCFG/EnforceSingleExitPass.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+      // If there is at least one `TrivialExit`, we use the first one as the
+      // `OneTrueExit`
+      OneTrueExit = *TrivialExits.begin();
+    }
+
+    // 7: Connect all the `TrivialExits`, except the first one, to the
+    // `OneTrueExit`
+    for (const auto &Group : enumerate(TrivialExits)) {
+      if (Group.index() < 1)
+        continue;
```
[the remark this answers is not in the document]
Comment (andrealinux1) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: Oh yeah, didn't think of it, point is `skip_front` assumes the range contains at least the elements to be skipped, so it needs to be guarded for the case we have no `TrivialExit`, so not super elegant too in this fashion. Anyway changes applied here https://github.com/revng/revng/blob/1a35abb5e36870643588163d5b6f0c0c72feb701/lib/RestructureCFG/EnforceSingleExitPass.cpp#L157

File: lib/RestructureCFG/EnforceSingleExitPass.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (pfez) [reply lib/RestructureCFG/EnforceSingleExitPass.cpp]: > I guess you meant to return true if you reach the end of the function

Yeah.

Thank you.

**Pull request merged**